src/checkpointing.jl

```diff
@@ -0,0 +1,29 @@
+"""
+    checkpoint(f, args::Tuple)
+
+Treat `f` as a primitive, whose gradient is computed by running and forwards- and reverse-
+pass of ∇ inside the outer pass. This can have the effect of significantly lowering the
+peak memory usage of a gradient computation at the expense of additional computation.
+
+HEALTH WARNING: If you close over differentiable objects in `f`, you will get incorrect
+results. It is therefore very much recommended to verify your gradients using finite
+differencing whenever code is checkpointed.
+"""
+checkpoint(f, args::Tuple) = f(args...)
```

Comment:
So one idea for handling keyword arguments:
```suggestion
checkpoint(f, args::Tuple, kwargs::NamedTuple=NamedTuple()) = f(args...; kwargs...)
```
Then define the `∇` with an optional `kwargs` positional argument. Also
```julia
@explicit_intercepts checkpoint Tuple{Any, Tuple} [false, true]
@explicit_intercepts checkpoint Tuple{Any, Tuple, NamedTuple} [false, true, false]
```
And change the `f(args_...)` to `f(args_...; kwargs...)` in the `∇` method.

Reply:
That means `checkpoint` could be invoked as, e.g.
```julia
checkpoint(sum, (x,), (dims=:,))
```